source/MRMesh/MRSharedThreadSafeOwner.cpp

```diff
@@ -0,0 +1,102 @@
+#include "MRSharedThreadSafeOwner.h"
+#include "MRAABBTree.h"
+#include "MRAABBTreePolyline.h"
+#include "MRAABBTreePoints.h"
+#include "MRDipole.h"
+#include "MRHeapBytes.h"
+#include "MRPch/MRTBB.h"
+#include <cassert>
+
+namespace MR
+{
+
+struct TaskGroup : tbb::task_group
+{};
+
+template<typename T>
+void SharedThreadSafeOwner<T>::reset()
+{
+    assert( !construction_ ); // one thread constructs the object, and this thread resets it
+    obj_.reset();
+}
+
+template<typename T>
+void SharedThreadSafeOwner<T>::update( const std::function<void(T&)> & updater )
+{
+    assert( !construction_ ); // one thread constructs the object, and this thread updates it
+    auto myPtr = atomic_exchange( &obj_, {} );
+    if ( !myPtr )
+        return;
+    assert( myPtr.use_count() >= 1 );
+    if ( myPtr.use_count() > 1 ) // create the copy if we was not the unique owner
+        myPtr.reset( new T( *myPtr ) );
+    assert( myPtr.use_count() == 1 );
+    updater( *myPtr );
+    assert( !construction_ ); // one thread constructs the object, and this thread updates it
+    atomic_store( &obj_, std::move( myPtr ) );
+}
+
+template<typename T>
+const T & SharedThreadSafeOwner<T>::getOrCreate( const std::function<T()> & creator )
+{
+    assert( creator );
+    /// if many parallel threads call this function simultaneously, they will join one task_group
+    /// and will cooperatively construct owned object
+    for (;;)
+    {
+        if ( auto p = obj_.get() ) // fastest path avoiding increasing/decreasing reference counter when everything is ready
+            return *p;
+        if ( auto p = atomic_load( &obj_ ) ) // fast path, which insures that the thread does not use old register-cached value
+            return *p;
+
+
+        bool firstConstructor = false; // true only the thread starting the creation
+        auto construction = atomic_load( &construction_ );
+        if ( !construction )
+        {
+            firstConstructor = atomic_compare_exchange_strong( &construction_, &construction, std::make_shared<TaskGroup>() );
+            if ( firstConstructor )
+                construction = atomic_load( &construction_ );
+            assert( construction );
+        }
+
+        if ( auto p = atomic_load( &obj_ ) ) // already constructed while we setup construction
+        {
+            if ( firstConstructor )
```

Comment:
Could `firstConstructor` ever be `true` in this branch?

Reply:
It is very unlikely, indeed. But it can happen, e.g. if update() method just finished its operation in another thread.